larpandoracontent/LArReclustering/ThreeDMultiReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,360 @@
+        return STATUS_CODE_SUCCESS;
+
+    // Ask the FOM alg tool which pfos to recluster
+    PfoList pfosToRecluster;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, m_pFomAlgTool->GetPfosToRecluster(pPfos, pfosToRecluster));
+    if (pfosToRecluster.empty())
+        return STATUS_CODE_SUCCESS;
+
+    // Remove clusters from the pfos, keeping track of their original pfo
+    std::map<HitType, ClusterList> viewToFreeClusters = { {TPC_3D, {}}, {TPC_VIEW_U, {}}, {TPC_VIEW_V, {}}, {TPC_VIEW_W, {}} };
+    std::map<const Pfo *const, ClusterList> pfoToFreeClusters;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, FreeClustersFromPfos(pfosToRecluster, viewToFreeClusters, pfoToFreeClusters));
+    if (viewToFreeClusters.at(TPC_3D).empty())
+        return STATUS_CODE_SUCCESS;
+
+    // Run reclustering algs over the free 3D clusters to see if there is a superior 3D clustering
+    // NOTE Temporarily replacing the current list doesn't seem to work
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ReplaceCurrentList<Cluster>(*this, m_cluster3DListName));
+    std::string originalClusterListName;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=,
+                             PandoraContentApi::InitializeReclustering(*this, TrackList(), viewToFreeClusters.at(TPC_3D), originalClusterListName));
+
+    float bestReclusterFom;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, m_pFomAlgTool->CalcClusteringFom(viewToFreeClusters.at(TPC_3D), bestReclusterFom));
+    std::string bestReclusterListName {originalClusterListName};
+    ClusterList bestReclusterList;
+
+    for (const std::string &clusteringAlg : m_clusteringAlgs)
+    {
+        std::string reclusterListName;
+        const ClusterList *pReclusterList {nullptr};
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=,
+                                 PandoraContentApi::RunClusteringAlgorithm(*this, clusteringAlg, pReclusterList, reclusterListName));
+        if (!pReclusterList || pReclusterList->empty())
+            continue;
+
+        float reclusterFom;
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, m_pFomAlgTool->CalcClusteringFom(*pReclusterList, reclusterFom));
+        if (reclusterFom > bestReclusterFom)
+        {
+            bestReclusterFom = reclusterFom;
+            bestReclusterListName = reclusterListName;
+            bestReclusterList = *pReclusterList;
+        }
+    }
+
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::TemporarilyReplaceCurrentList<Cluster>(*this, bestReclusterListName));
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::EndReclustering(*this, bestReclusterListName));
+
+    if (bestReclusterListName == originalClusterListName) // Return the clusters to the original pfos as if nothing happened...
+    {
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, AddClustersToPfos(pfoToFreeClusters));
+        return STATUS_CODE_SUCCESS;
```

Comment:
could you not just do `return AddClustersToPfos(pfoToFreeClusters) `

Reply:
changed, thanks!